x/gamm/keeper/multihop_test.go

```diff
@@ -68,33 +102,48 @@ func (suite *KeeperTestSuite) TestBalancerPoolSimpleMultihopSwapExactAmountIn()
 						dec = dec.Mul(sp)
 					}
 					return dec
-				}()
+				}
 
-				tokenOutAmount, err := keeper.MultihopSwapExactAmountIn(suite.Ctx, suite.TestAccs[0], test.param.routes, test.param.tokenIn, test.param.tokenOutMinAmount)
-				suite.NoError(err, "test: %v", test.name)
+				// we create exact the same except swap fee 2 pool pairs
+				// use +2 to calc using second pair (w/o swap fee)
+				calcOutAmountAsSeparateSwaps := func(poolIdShift uint64) sdk.Coin {
+					cacheCtx, _ := suite.Ctx.CacheContext()
+					nextTokenIn := test.param.tokenIn
+					for _, hop := range test.param.routes {
 
-				// Calculate the chained spot price.
-				spotPriceAfter := func() sdk.Dec {
-					dec := sdk.NewDec(1)
-					tokenInDenom := test.param.tokenIn.Denom
-					for i, route := range test.param.routes {
-						if i != 0 {
-							tokenInDenom = test.param.routes[i-1].TokenOutDenom
-						}
+						tokenOut, err := keeper.SwapExactAmountIn(cacheCtx, suite.TestAccs[0], hop.PoolId+poolIdShift, nextTokenIn, hop.TokenOutDenom, sdk.OneInt())
+						suite.Require().NoError(err)
+						nextTokenIn = sdk.NewCoin(hop.TokenOutDenom, tokenOut)
+					}
+					return nextTokenIn
+				}
 
-						pool, err := keeper.GetPoolAndPoke(suite.Ctx, route.PoolId)
-						suite.NoError(err, "test: %v", test.name)
+				tokenOutCalculatedAsSeparateSwaps := calcOutAmountAsSeparateSwaps(0)
+				tokenOutCalculatedAsSeparateSwapsWithoutFee := calcOutAmountAsSeparateSwaps(2)
 
-						sp, err := pool.SpotPrice(suite.Ctx, tokenInDenom, route.TokenOutDenom)
-						suite.NoError(err, "test: %v", test.name)
-						dec = dec.Mul(sp)
-					}
-					return dec
-				}()
+				spotPriceBefore := calcSpotPrice()
+
+				tokenOutAmount, err := keeper.MultihopSwapExactAmountIn(suite.Ctx, suite.TestAccs[0], test.param.routes, test.param.tokenIn, test.param.tokenOutMinAmount)
+				suite.NoError(err, "test: %v", test.name)
+
+				spotPriceAfter := calcSpotPrice()
 
 				// Ratio of the token out should be between the before spot price and after spot price.
 				sp := test.param.tokenIn.Amount.ToDec().Quo(tokenOutAmount.ToDec())
 				suite.True(sp.GT(spotPriceBefore) && sp.LT(spotPriceAfter), "test: %v", test.name)
+
+				if test.reducedFeeApplied {
+					// here we do not have exact 50% reduce due to amm math rounding and other staff
+					// playing with input values for this test can result in different discount %
+					// so lets check, that we have around 50% +-1% reduction
+					diffA := tokenOutCalculatedAsSeparateSwapsWithoutFee.Amount.Sub(tokenOutAmount)
+					diffB := tokenOutAmount.Sub(tokenOutCalculatedAsSeparateSwaps.Amount)
+					diffDistinctionPercent := diffA.Sub(diffB).Abs().ToDec().Quo(diffA.Add(diffB).ToDec())
+					suite.Require().True(diffDistinctionPercent.LT(sdk.MustNewDecFromStr("0.01")))
```

Comment:
Would like to do some work here to make this slightly clearer, such as:

1) for diffB, subtract in the same order as we did for diffA (then add the differences rather than subtract)
2) add comments explaining what exactly diffA and diffB represent